sys/usb/usbus/usbus_control.c

```diff
@@ -0,0 +1,408 @@
+                }
+            }
+        }
+    }
+}
+
+static size_t usbus_cpy_str_to_utf16(usbus_t *usbus, const char *str)
+{
+    size_t len = 0;
+
+    while (*str) {
+        usbus_control_slicer_put_char(usbus, *str);
+        usbus_control_slicer_put_char(usbus, 0);
+        len += 2; /* Two bytes added each iteration */
+        str++;
+    }
+    return len;
+}
+
+static usbus_string_t *_get_descriptor(usbus_t *usbus, uint16_t idx)
+{
+    for (usbus_string_t *str = usbus->strings; str; str = str->next) {
+        if (str->idx == idx) {
+            return str;
+        }
+    }
+    return NULL;
+}
+
+static int _req_status(usbus_t *usbus)
+{
+    uint8_t status[2];
+
+    memset(status, 0, sizeof(status));
+    usbus_control_slicer_put_bytes(usbus, status, sizeof(status));
+    return sizeof(status);
+}
+
+static int _req_str(usbus_t *usbus, uint16_t idx)
+{
+    if (idx == 0) {
+        usb_descriptor_string_t desc;
+        desc.type = USB_TYPE_DESCRIPTOR_STRING;
+        desc.length = sizeof(uint16_t) + sizeof(usb_descriptor_string_t);
+        usbus_control_slicer_put_bytes(usbus, (uint8_t *)&desc, sizeof(desc));
+        /* Only one language ID supported */
+        uint16_t us = USB_CONFIG_DEFAULT_LANGID;
+        usbus_control_slicer_put_bytes(usbus, (uint8_t *)&us, sizeof(uint16_t));
+    }
+    else {
+        usb_descriptor_string_t desc;
+        desc.type = USB_TYPE_DESCRIPTOR_STRING;
+        usbus_string_t *str = _get_descriptor(usbus, idx);
+        if (str) {
+            desc.length = sizeof(usb_descriptor_string_t);
+            desc.length += 2 * strlen(str->str); /* USB strings are UTF-16 */
+            usbus_control_slicer_put_bytes(usbus, (uint8_t *)&desc,
+                                           sizeof(desc));
+            usbus_cpy_str_to_utf16(usbus, str->str);
+        }
+    }
+    return 1;
+}
+
+static int _req_dev(usbus_t *usbus)
+{
+    return usbus_fmt_hdr_dev(usbus);
+}
+
+static int _req_config(usbus_t *usbus)
+{
+    usbus_fmt_hdr_conf(usbus);
+    return 1;
+}
+
+static int _req_dev_qualifier(usbus_t *usbus)
+{
+    usb_speed_t speed = USB_SPEED_LOW;
+
+    usbus->dev->driver->get(usbus->dev, USBOPT_MAX_SPEED, &speed,
+                            sizeof(usb_speed_t));
+    if (speed == USB_SPEED_HIGH) {
+        /* TODO: implement device qualifier support (only required
+         * for High speed) */
+    }
+    return -1;
+}
+
+static int _req_descriptor(usbus_t *usbus, usb_setup_t *pkt)
+{
+    uint8_t type = pkt->value >> 8;
+    uint8_t idx = (uint8_t)pkt->value;
+
+    /* Decode descriptor type */
+    switch (type) {
+        case USB_TYPE_DESCRIPTOR_DEVICE:
+            return _req_dev(usbus);
+        case USB_TYPE_DESCRIPTOR_CONFIGURATION:
+            return _req_config(usbus);
+        case USB_TYPE_DESCRIPTOR_STRING:
+            return _req_str(usbus, idx);
+        case USB_TYPE_DESCRIPTOR_DEV_QUALIFIER:
+            return _req_dev_qualifier(usbus);
+        default:
```

Comment:
please add a DEBUG statement here and in other default statements below.

Reply:
I think I got them all, but feel free to point out if I missed some